emu_base/pulser_adapter.py

```diff
@@ -117,34 +120,63 @@ def _extract_omega_delta_phi(
     nsamples = len(target_times) - 1
     omega = torch.zeros(
         nsamples,
-        len(sequence.register.qubit_ids),
+        len(q_ids),
         dtype=torch.complex128,
     )
 
     delta = torch.zeros(
         nsamples,
-        len(sequence.register.qubit_ids),
+        len(q_ids),
         dtype=torch.complex128,
     )
     phi = torch.zeros(
         nsamples,
-        len(sequence.register.qubit_ids),
+        len(q_ids),
         dtype=torch.complex128,
     )
+    qubit_positions = _get_qubit_positions(sequence.register)
 
-    if laser_waist:
-        qubit_positions = _get_qubit_positions(sequence.register)
-        waist_factors = torch.tensor(
-            [math.exp(-((x[:2].norm() / laser_waist) ** 2)) for x in qubit_positions]
-        )
-    else:
-        waist_factors = torch.ones(len(sequence.register.qubit_ids))
+    def perp_dist(pos: torch.Tensor, axis: torch.Tensor) -> Any:
+        return torch.linalg.vector_norm(pos - torch.vdot(pos, axis) * axis)
 
-    global_times = set()
+    times_to_amp_factors: dict[int, torch.Tensor] = {}
     for ch, ch_samples in samples.channel_samples.items():
-        if samples._ch_objs[ch].addressing == "Global":
-            for slot in ch_samples.slots:
-                global_times |= set(i for i in range(slot.ti, slot.tf))
+        ch_obj = samples._ch_objs[ch]
+        prop_dir = torch.tensor(
+            ch_obj.propagation_dir or [0.0, 1.0, 0.0], dtype=torch.float64
+        )
+        prop_dir /= prop_dir.norm()
+
+        # each channel has a noise on its laser amplitude
+        # we assume each channel has the same noise amplitude currently
+        # the hardware currently has only a global channel anyway
+        sigma_factor = torch.max(
```

Comment:
I find a bit sad that you had to modify bitstring counts on non-noisy end-to-end tests. Maybe to avoid changing the state of the random number generator, I personally wouldn't be shocked by
```
sigma_factor = 1.0 if amp_sigma == 0.0 else max(
            0, 1.0 + amp_sigma**2 * torch.randn(1).item()
        )
```

It's also a sanity test - if the bitstrings do not go back to normal then we have another problem?

Reply:
good point. I also realized during my sleep that the scale factor should be `amp_sigma` not the square.